tools/regtesting/do_regtest.py

```diff
@@ -0,0 +1,437 @@
+    timings = sorted([r.duration for r in all_results])
+    print('Plot: name="timings", title="Timing Distribution", ylabel="time [s]"')
+    for p in (100, 99, 98, 95, 90, 80):
+        v = percentile(timings, p / 100.0)
+        print(f'PlotPoint: name="{p}th_percentile", plot="timings", ', end="")
+        print(f'label="{p}th %ile", y={v:.2f}, yerr=0.0')
+
+    print("\n-------------------------------- Summary -------------------------------")
+    total_duration = time.perf_counter() - start_time
+    num_tests = len(all_results)
+    num_failed = sum([r.status in ("TIMEOUT", "RUNTIME FAIL") for r in all_results])
+    num_wrong = sum([r.status == "WRONG RESULT" for r in all_results])
+    num_ok = sum([r.status == "OK" for r in all_results])
+    print(f"Number of FAILED  tests {num_failed}")
+    print(f"Number of WRONG   tests {num_wrong}")
+    print(f"Number of CORRECT tests {num_ok}")
+    print(f"Total number of   tests {num_tests}")
+    summary = f"\nSummary: correct: {num_ok} / {num_tests}"
+    summary += f"; wrong: {num_wrong}" if num_wrong > 0 else ""
+    summary += f"; failed: {num_failed}" if num_failed > 0 else ""
+    summary += f"; {total_duration/60.0:.0f}min"
+    print(summary)
+    print("Status: " + ("OK" if num_ok == num_tests else "FAILED") + "\n")
+
+    print("*************************** testing ended ******************************")
+    sys.exit(num_tests - num_ok)
+
+
+# ======================================================================================
+class Config:
+    def __init__(self, args: argparse.Namespace):
+        self.timeout = args.timeout
+        self.use_mpi = args.version.startswith("p")
+        self.ompthreads = args.ompthreads
+        self.mpiranks = args.mpiranks if self.use_mpi else 1
+        self.num_workers = int(args.maxtasks / self.ompthreads / self.mpiranks)
+        self.workers = Semaphore(self.num_workers)
+        self.cp2k_root = Path(__file__).resolve().parent.parent.parent
+        datestamp = datetime.now().strftime("%Y-%m-%d_%H-%M-%S")
+        leaf_dir = f"TEST-{args.arch}-{args.version}-{datestamp}"
+        self.work_base_dir = self.cp2k_root / "regtesting" / leaf_dir
+        self.error_summary = self.work_base_dir / "error_summary"
+        self.keep_alive = args.keep_alive
+        self.arch = args.arch
+        self.version = args.version
+        self.debug = args.debug
+        self.restrictdirs = args.restrictdir if args.restrictdir else [".*"]
+        self.gpu_round_robin = 0
+        self.num_gpu_devices = 2  # TODO detect number of GPUs
+
+    def cp2k_command(self) -> List[str]:
+        if self.mpiranks == 1:  # Speedup non-mpi test by using all available GPUs.
+            self.gpu_round_robin = (self.gpu_round_robin + 1) % self.num_gpu_devices
+            os.environ["CUDA_VISIBLE_DEVICES"] = str(self.gpu_round_robin)
+        os.environ["OMP_NUM_THREADS"] = str(self.ompthreads)
```

Comment:
I'd advise against manipulating the current environ (especially when you're delaying the actual process making use of that environ) and rather pass an explicit dict with the `env` keyword to wherever you're spawning the process where the dict passing is a merge of `os.environ` and what `cp2k_command` should return. So the signature would then rather be:

```
     def cp2k_command(self) -> Tuple(List[str], Dict[str,str]):
```

Reply:
Good point. I refactor the launching of binaries.